rs/nns/governance/src/node_provider_rewards.rs

```diff
@@ -30,17 +47,53 @@ pub(crate) fn latest_node_provider_rewards() -> Option<ArchivedMonthlyNodeProvid
     })
 }
 
-pub(crate) fn list_node_provider_rewards(limit: u64) -> Vec<ArchivedMonthlyNodeProviderRewards> {
-    // limit try_into is safe because of u32 is always equal or smaller than usize
+pub(crate) fn list_node_provider_rewards(
+    limit: u64,
+    date_filter: Option<DateRangeFilter>,
+) -> Vec<ArchivedMonthlyNodeProviderRewards> {
+    let date_filter = date_filter.unwrap_or_default();
+    let start_timestamp = date_filter.start.unwrap_or(0);
+    let end_timestamp = date_filter.end.unwrap_or(u64::MAX);
+
     with_node_provider_rewards_log(|log| {
-        let len: u64 = log.len();
+        // naive filtering implementation is okay b/c our data set is very small
+        // (1/month for years will not give us much to work through for some time)
+        // TODO - migrate to BTreeMap storage with keys as dates and change this to use range lookup
+        let rewards = log
+            .iter()
+            .flat_map(|rewards| {
+                let timestamp = rewards
+                    .version
+                    .clone()
+                    .map(|v| match v {
+                        Version::Version1(v1) => v1,
+                    })
+                    .and_then(|v1| v1.rewards)
+                    .map(|rewards| rewards.timestamp);
+
+                if let Some(timestamp) = timestamp {
+                    if timestamp >= start_timestamp && timestamp <= end_timestamp {
+                        Some(rewards)
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let len: u64 = rewards.len().try_into().unwrap();
 
         let end_range = len;
+        // start of range is just the end minus the limit, but not less than 0
         let start_range = end_range.saturating_sub(limit);
 
+        // If we have 10 entries, they're 0..9
+        // If we are getting newest first, we want to return 9..4, then 4..0, thus the rev() call
         (start_range..end_range)
             .rev()
-            .flat_map(|index| log.get(index))
+            .map(|index| rewards[index as usize].clone())
             .collect()
```

Comment:
Can this be achieved with `rewards.into_iter().rev().take(limit)`?

Reply:
Yes.  This was leftover from originally trying with log's iterator, which couldn't be reversed.  Good catch.